storage/innobase/lock/lock0lock.cc

```diff
@@ -573,8 +557,83 @@ static void wsrep_assert_no_bf_bf_wait(const lock_t *lock, const trx_t *trx,
 	/* BF-BF wait is a bug */
 	ut_error;
 }
+
+void wsrep_report_error(const lock_t* victim_lock, const trx_t *bf_trx);
 # endif /* UNIV_DEBUG */
 
+/** Check if a high priority (BF) trx has to wait for the current
+lock holder based on Wsrep transaction state relations.
+
+This code resembles the one in `wsrep_handle_mdl_conflict()`, but
+it's specific to the storage engine and it doesn't perform any
+BF aborts by itself.
+The decision whether to BF abort a victim may depend on other conditions
+like lock compatibility between InnoDB transactions.
+
+@param lock  other waiting lock
+@param trx   BF transaction requesting conflicting lock
+*/
+static bool wsrep_BF_has_to_wait(const lock_t *lock, const trx_t *trx,
+                                 bool report_bf_bf_wait,
+                                 const unsigned type_mode = LOCK_NONE)
+{
+  THD *request_thd= trx->mysql_thd;
+  THD *granted_thd= lock->trx->mysql_thd;
+
+  /* Assume requester is BF. */
+  ut_ad(wsrep_thd_is_BF(request_thd, false));
+
+  /* Assume lock holder is Wsrep. */
+  ut_ad(lock->trx->is_wsrep());
+
+  /* Granted is aborting, let it complete. */
+  if (wsrep_thd_is_aborting(granted_thd))
+    return true;
+
+  /* Granted is not BF, may BF abort it. */
+  if (!wsrep_thd_is_BF(granted_thd, false))
+    return false;
+
+  /* Applying SR cannot BF abort other high priority (BF). */
+  if (wsrep_thd_is_SR(request_thd))
+    return true;
+
+  /* Requester is truly BF and granted is applying SR in progress. */
+  if (wsrep_thd_is_SR(granted_thd))
+    return false;
+
+#ifdef UNIV_DEBUG
+  if (report_bf_bf_wait)
+    wsrep_report_error(lock, trx);
+  /* We very well can let BF to wait normally as other
+     BF will be replayed in case of conflict. For debug
+     builds we will do additional sanity checks to catch
+     unsupported BF wait if any. */
+  ut_d(wsrep_assert_valid_bf_bf_wait(lock, trx, type_mode));
+#endif
+  return true;
+}
+
+/** Determine whether BF abort on the lock holder is needed.
+
+@param lock  other waiting lock
+@param trx   BF transaction requesting conflicting lock
+*/
+static bool wsrep_will_BF_abort(const lock_t *lock, const trx_t *trx)
+{
+  /* Assume requester is BF. */
+  ut_ad(wsrep_thd_is_BF(trx->mysql_thd, false));
+
+  /* Lock holder is not BF - can BF abort it. */
+  if (!wsrep_thd_is_BF(lock->trx->mysql_thd, false))
```

Comment:
Yes, it is. Frankly, I put it here because I don't know if it's possible to see lock->trx->is_wsrep() ==  false in this code, so I put an assert in wsrep_BF_has_to_wait() instead and filter non-wsrep trxs here as part of this check, just because almost the same check existed before. Do you know if non-wsrep transactions are possible and if we should BF abort them?

Reply:
I don't expect to see non-wsrep transactions here (maybe a debug assertion would be in place at the beginning of the function). However, wsrep transactions which are *not BF* will always return here, without a lock wait check.

The same check seems to exist in `wsrep_BF_has_to_wait()`, maybe it would be enough?

Reply:
Yes, I will try an assert instead.